### DISCUSSION TITLE
ms_token

### DISCUSSION THREAD

**Discussion by kulig1985** [davidteather/TikTok-Api#1101] [General]
how to get the proper ms token? or wher and how to generate it? Thanks!

**Comment by jesse-moderwell** [comment]
I was wondering this myself. Here's how I'm doing it so far:

- Go to tiktok.com
- right click then do inspect 
- find the application tab (for me I have to click on the arrows >>)
- Then click on cookies, and there should be a bunch of values in a table of sorts. There are several items marked 'msToken'

Not sure if that is technically the correct way. It does seem to work some times, but I'm running into other issues. Hope that helps

**Reply by Beedrick** [reply]
Tried this and am getting errors as follows:

Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "C:\Users\Gorilla Rig\Desktop\RnD\TiktokWebscraper\TikTok-Api\examples\trending_example.py", line 20, in <module>
    asyncio.run(trending_videos())
  File "C:\Python311\Lib\asyncio\runners.py", line 190, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "C:\Python311\Lib\asyncio\runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Python311\Lib\asyncio\base_events.py", line 653, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "C:\Users\Gorilla Rig\Desktop\RnD\TiktokWebscraper\TikTok-Api\examples\trending_example.py", line 11, in trending_videos
    async for video in api.trending.videos(count=30):
  File "C:\Users\Gorilla Rig\Desktop\RnD\TiktokWebscraper\TikTok-Api\TikTokApi\api\trending.py", line 43, in videos
    resp = await Trending.parent.make_request(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Gorilla Rig\Desktop\RnD\TiktokWebscraper\TikTok-Api\TikTokApi\tiktok.py", line 441, in make_request
    raise EmptyResponseException(result, "TikTok returned an empty response")
TikTokApi.exceptions.EmptyResponseException: None -> TikTok returned an empty response

any update on this issue? I am trying to run the trending_example.py file using python -m examples.trending_examples but its failing to retrieve any data

**Comment by zakrian07** [comment]
Help me too how to pick correct ms token

**Comment by ActualMasterOogway** [comment]
Yes its kinda weird because my msToken keeps changing every minute or so

**Comment by sebastianfeistl** [comment]
@ActualMasterOogway 

I can confirm the `ms_token` changes frequently, sometimes even every second. By the time I run my script, the token may have expired, resulting in no output. Despite meeting all prerequisites, I haven't been able to run a single successful example using this repository.

The documentation, including the linked YouTube video, seems outdated and causes confusion. Without proper documentation, this repository is not reliable. I recommend finding a different API for TikTok integration.

**Reply by jianhuayang** [reply]
same here, is there a solution?

**Reply by rahults2016** [reply]
Does anybody find a solution to this? I can see ms_token changing very frequently
@jianhuayang @ActualMasterOogway

**Reply by jianhuayang** [reply]
@rahults2016 @ActualMasterOogway 
what I figured out was 1) don't use any existing msToken, let the API generate for you, and 2) add below after `await page.goto(url)`

```
        await page.goto(url)
        await page.wait_for_timeout(3000)
        await page.click("button:text('Accept All')")
        await page.wait_for_timeout(3000)
```
this seemed to be working for me.

**Comment by louissiu198** [comment]
Are you guys dumb, msToken is being recreated every request from the response cookie, but the only valid msToken is the one which is using the common / report endpoint by encrypting the strData (web mssdk), the the strData is the one which depends if its valid or not

**Reply by tarik0** [reply]
your profile pic fits this comment perfectly
